python/turbine_models/custom_models/sd_inference/unet_test.py

```diff
@@ -0,0 +1,247 @@
+import torch
+import torch._dynamo as dynamo
+from diffusers import UNet2DConditionModel
+
+import safetensors
+import argparse
+
+parser = argparse.ArgumentParser()
+parser.add_argument(
+    "--hf_auth_token", type=str, help="The Hugging Face auth token, required"
+)
+parser.add_argument(
+    "--hf_model_name",
+    type=str,
+    help="HF model name",
+    default="CompVis/stable-diffusion-v1-4",
+)
+parser.add_argument("--run_vmfb", action="store_true")
+parser.add_argument("--compile_to", type=str, help="torch, linalg, vmfb")
+parser.add_argument("--external_weight_file", type=str, default="")
+parser.add_argument("--vmfb_path", type=str, default="")
+parser.add_argument(
+    "--external_weights",
+    type=str,
+    default=None,
+    help="saves ir/vmfb without global weights for size and readability, options [safetensors]",
+)
+
+
+class UnetModel(torch.nn.Module):
+    def __init__(self, args):
+        super().__init__()
+        self.unet = UNet2DConditionModel.from_pretrained(
+            args.hf_model_name,
+            subfolder="unet",
+            token=args.hf_auth_token,
+        )
+        self.guidance_scale = 7.5
+
+    def forward(self, sample, timestep, encoder_hidden_states):
+        samples = torch.cat([sample] * 2)
+        unet_out = self.unet.forward(
+            samples, timestep, encoder_hidden_states, return_dict=False
+        )[0]
+        noise_pred_uncond, noise_pred_text = unet_out.chunk(2)
+        noise_pred = noise_pred_uncond + self.guidance_scale * (
+            noise_pred_text - noise_pred_uncond
+        )
+        return noise_pred
+
+
+def save_external_weights(
+    mapper,
+    model,
+    external_weights=None,
+    external_weight_file=None,
+):
+    if external_weights is not None:
+        if external_weights == "safetensors":
+            mod_params = dict(model.named_parameters())
+            for name in mod_params:
+                mapper["params." + name] = name
+            if external_weight_file:
+                safetensors.torch.save_file(mod_params, external_weight_file)
+                print("Saved params to", external_weight_file)
+
+
+def export_unet_model(args, unet_model):
+    mapper = {}
+    save_external_weights(
+        mapper, unet_model, args.external_weights, args.external_weight_file
+    )
+
+    if args.hf_model_name == "CompVis/stable-diffusion-v1-4":
+
+        class CompiledUnet(CompiledModule):
+            if args.external_weights:
+                params = export_parameters(
+                    unet_model, external=True, external_scope="", name_mapper=mapper.get
+                )
+            else:
+                params = export_parameters(unet_model)
+
+            def main(
+                self,
+                sample=AbstractTensor(1, 4, 64, 64, dtype=torch.float32),
+                timestep=AbstractTensor(1, dtype=torch.float32),
+                encoder_hidden_states=AbstractTensor(2, 77, 768, dtype=torch.float32),
```

Comment:
need to find better way to change AbstractTensor size

Reply:
What do you mean? You can use None for dynamic if needed

Reply:
Hey @dan-garvey, the issue is that for SD 1.4/2.1 the last dimension of encoder_hidden_states is 768/1024, respectively. When I try to use None in the AbstractTensor with the constraint set to `encoder_hidden_states.dynamic_dim(2)`, I get this error:

RuntimeError: a and b must have same reduction dim, but got [s2*s3, s2] X [768, 320].

Reply:
I see, yeah unless the encoder hidden state is also dynamic it won't work. In this case I'd just parameterize the last dim based on which model you're instantiating